Support Kotlin unsigned integer types

你好！现在在 entity 中使用 Kotlin 的两个无符号整型（`UInt`, `ULong`）会出现异常，无法正常使用。

**Entity** :

```kotlin
@Entity
interface Building {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    val id: Long // `id bigserial`

    @Key
    val name: String

    val value: UInt
}
```

**Unit Test** 

```kotlin
@SpringBootTest(webEnvironment = MOCK)
@ActiveProfiles("...")
@AutoConfigureTestDatabase(replace = NONE)
class BuildingTests {
    @Autowired
    lateinit var sqlClient: KSqlClient

    @Test
    @Transactional
    @Rollback
    fun testBuild() {
        val new = Building {
            name = "New Building"
            value = 50u
        }

        val saved = sqlClient.save(new)
            .modifiedEntity

        println(saved)
    }
}
```

**Log**

```
Internal bug: Cannot find the getter of prop "value" of the interface "org.faraday.rs.entity.Building"
java.lang.AssertionError: Internal bug: Cannot find the getter of prop "value" of the interface "org.faraday.rs.entity.Building"
	at org.babyfish.jimmer.meta.impl.ImmutablePropImpl.<init>(ImmutablePropImpl.java:199)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$PropBuilder.build(ImmutableTypeImpl.java:1235)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$BuilderImpl.build(ImmutableTypeImpl.java:1160)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$BuilderImpl.build(ImmutableTypeImpl.java:621)
	at org.faraday.rs.entity.BuildingDraft$$.<clinit>(BuildingDraft.kt:70)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/java.lang.reflect.Field.acquireFieldAccessor(Field.java:1185)
	at java.base/java.lang.reflect.Field.getFieldAccessor(Field.java:1164)
	at java.base/java.lang.reflect.Field.get(Field.java:442)
	at org.babyfish.jimmer.meta.impl.Metadata.create(Metadata.java:90)
	at org.babyfish.jimmer.impl.util.StaticCache.getWithoutLock(StaticCache.java:68)
	at org.babyfish.jimmer.impl.util.StaticCache.get(StaticCache.java:54)
	at org.babyfish.jimmer.meta.impl.Metadata.get(Metadata.java:28)
	at org.babyfish.jimmer.meta.ImmutableType.get(ImmutableType.java:26)
	at org.babyfish.jimmer.sql.runtime.EntityManager.<init>(EntityManager.java:62)
	at org.babyfish.jimmer.sql.runtime.EntityManager.fromResources(EntityManager.java:232)
	at org.babyfish.jimmer.sql.JSqlClientImpl$BuilderImpl.entityManager(JSqlClientImpl.java:1782)
	at org.babyfish.jimmer.sql.JSqlClientImpl$BuilderImpl.build(JSqlClientImpl.java:1573)
	at org.babyfish.jimmer.sql.di.JLazyInitializationSqlClient.sqlClient(JLazyInitializationSqlClient.java:35)
	at org.babyfish.jimmer.sql.di.AbstractJSqlClientDelegate.getEntities(AbstractJSqlClientDelegate.java:90)
	at org.babyfish.jimmer.sql.kt.impl.KSqlClientImpl.<init>(KSqlClientImpl.kt:76)
	at org.babyfish.jimmer.sql.kt.KSqlClientKt.toKSqlClient(KSqlClient.kt:1068)
	at org.babyfish.jimmer.spring.SqlClients.kotlin(SqlClients.kt:74)
	at org.babyfish.jimmer.spring.SqlClients.kotlin(SqlClients.kt:43)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig.kotlinSqlClient(SqlClientConfig.java:31)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$0.CGLIB$kotlinSqlClient$1(<generated>)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$FastClass$$1.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:348)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$0.kotlinSqlClient(<generated>)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:644)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:636)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:975)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:971)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
	at org.springframework.boot.test.context.SpringBootContextLoader.lambda$loadContext$3(SpringBootContextLoader.java:137)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:58)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:46)
	at org.springframework.boot.SpringApplication.withHook(SpringApplication.java:1463)
	at org.springframework.boot.test.context.SpringBootContextLoader$ContextLoaderHook.run(SpringBootContextLoader.java:553)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:137)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:108)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:225)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:152)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:130)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:191)
	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:130)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:260)
	at org.springframework.test.context.junit.jupiter.SpringExtension.postProcessTestInstance(SpringExtension.java:163)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

简单调试发现: https://github.com/babyfish-ct/jimmer/blob/f8f768192338144cf6920f25cd2ded7f47011773/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/impl/ImmutablePropImpl.java#L194-L202

此处由于 `UInt` 是 `value class`, 在Java中会被编译为 `getValue-pVg5ArA` 而不是 `getValue`, 因此无法被寻得。

_也许可以考虑当 `kotlinClass` 或 `kotlinProp` 存在时通过Kotlin而不是Java获取 getter?  _

并且一些运算操作符似乎也不支持，例如

```kotlin
sqlClient.createQuery(Building::class) {
            where(table.id.plus(3) eq 1) // 普通数字可以 plus
            // where(table.value.plus(3u) eq 1u) // 不支持 plus 操作符

           // 也许可以通过native sql临时处理，只不过因为上述错误暂时无法测试
            where(
                sql(UInt::class, "%e + %v") {
                    expression(table.value)
                    value(3u)
                }.eq(1u)
            )
            select(table)
        }.execute()
```

--- a/project/jimmer-core/build.gradle.kts
+++ b/project/jimmer-core/build.gradle.kts
@@ -8,6 +8,7 @@ dependencies {
     api(libs.jackson.databind)
     api(libs.kotlin.reflect)
     implementation(libs.jackson.datatype.jsr310)
+    implementation(libs.jackson.module.kotlin)
     implementation(libs.apache.commons.lang3)
     implementation(libs.kotlin.stdlib)
     compileOnly(libs.mapstruct)

--- a/project/jimmer-core/src/main/java/org/babyfish/jimmer/ImmutableObjects.java
+++ b/project/jimmer-core/src/main/java/org/babyfish/jimmer/ImmutableObjects.java
@@ -4,6 +4,7 @@ import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializationFeature;
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.fasterxml.jackson.module.kotlin.KotlinModule;
 import org.babyfish.jimmer.jackson.ImmutableModule;
 import org.babyfish.jimmer.meta.*;
 import org.babyfish.jimmer.runtime.DraftSpi;
@@ -483,6 +484,7 @@ public class ImmutableObjects {
         ObjectMapper mapper = new ObjectMapper();
         mapper.registerModule(new JavaTimeModule());
         mapper.registerModule(new ImmutableModule());
+        mapper.registerModule(new KotlinModule.Builder().build());
         mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
         MAPPER = mapper;
     }

--- a/project/jimmer-core/src/main/java/org/babyfish/jimmer/jackson/ImmutableProps.java
+++ b/project/jimmer-core/src/main/java/org/babyfish/jimmer/jackson/ImmutableProps.java
@@ -8,12 +8,16 @@ import java.lang.reflect.Method;
 
 class ImmutableProps {
 
-    private ImmutableProps() {}
+    private ImmutableProps() {
+    }
 
     static ImmutableProp get(ImmutableType type, Method method) {
         String propName = StringUtil.propName(method.getName(), false);
         if (propName == null) {
             propName = method.getName();
+        }
+        if (propName.matches(".*(-s-VKNKU|-pVg5ArA|-Mh2AYeg|-w2LRezQ)")) {
+            propName = propName.substring(0, propName.length() - 8);
         }
         ImmutableProp prop = type.getProps().get(propName);
         if (prop == null && (method.getReturnType() == boolean.class || method.getReturnType() == Boolean.class)) {

--- a/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/ImmutableProp.java
+++ b/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/ImmutableProp.java
@@ -1,5 +1,9 @@
 package org.babyfish.jimmer.meta;
 
+import kotlin.UByte;
+import kotlin.UInt;
+import kotlin.ULong;
+import kotlin.UShort;
 import org.babyfish.jimmer.jackson.Converter;
 import org.babyfish.jimmer.jackson.ConverterMetadata;
 import org.babyfish.jimmer.lang.Ref;
@@ -52,6 +56,14 @@ public interface ImmutableProp {
     boolean isInputNotNull();
 
     boolean isMutable();
+
+    default boolean isUnsigned() {
+        final String name = getElementClass().getName();
+        return name.equals(ULong.class.getName()) ||
+                name.equals(UInt.class.getName()) ||
+                name.equals(UShort.class.getName()) ||
+                name.equals(UByte.class.getName());
+    }
 
     <A extends Annotation> A getAnnotation(Class<A> annotationType);
 

--- a/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/impl/ImmutablePropImpl.java
+++ b/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/impl/ImmutablePropImpl.java
@@ -1,6 +1,10 @@
 package org.babyfish.jimmer.meta.impl;
 
 import com.fasterxml.jackson.annotation.JsonFormat;
+import kotlin.UByte;
+import kotlin.UInt;
+import kotlin.ULong;
+import kotlin.UShort;
 import kotlin.reflect.KClass;
 import kotlin.reflect.KProperty1;
 import kotlin.reflect.full.KClasses;
@@ -179,8 +183,19 @@ class ImmutablePropImpl implements ImmutableProp, ImmutablePropImplementor {
         } catch (NoSuchMethodException ignored) {
         }
         try {
-            javaGetter = declaringType.getJavaClass().getDeclaredMethod(
-                    "get" + name.substring(0, 1).toUpperCase() + name.substring(1));
+            String methodName = "get" + name.substring(0, 1).toUpperCase() + name.substring(1);
+            if (kotlinProp != null) {
+                if (elementClass.getName().equals(ULong.class.getName())) {
+                    methodName += "-s-VKNKU";
+                } else if (elementClass.getName().equals(UInt.class.getName())) {
+                    methodName += "-pVg5ArA";
+                } else if (elementClass.getName().equals(UShort.class.getName())) {
+                    methodName += "-Mh2AYeg";
+                } else if (elementClass.getName().equals(UByte.class.getName())) {
+                    methodName += "-w2LRezQ";
+                }
+            }
+            javaGetter = declaringType.getJavaClass().getDeclaredMethod(methodName);
         } catch (NoSuchMethodException ignored) {
         }
         try {

--- a/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/BuilderGenerator.kt
+++ b/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/BuilderGenerator.kt
@@ -84,7 +84,7 @@ class BuilderGenerator(
                             prop.slotName
                         )
                     } else {
-                        beginControlFlow("if (%L !== null)", prop.name)
+                        beginControlFlow("if (%L ${if (prop.isUnsigned) "!=" else "!=="} null)", prop.name)
                         addStatement("__draft.%L = %L", prop.name, prop.name)
                         addStatement(
                             "__draft.__show(%T.byIndex(%T.%L), true)",

--- a/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/DraftImplGenerator.kt
+++ b/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/DraftImplGenerator.kt
@@ -328,7 +328,7 @@ class DraftImplGenerator(
                             .apply {
                                 if (prop.isNullable) {
                                     beginControlFlow(
-                                        "if (!__isLoaded(%T.byIndex(%L)) || %L === null)",
+                                        "if (!__isLoaded(%T.byIndex(%L)) || %L ${if (prop.isUnsigned) "==" else "==="} null)",
                                         PROP_ID_CLASS_NAME,
                                         prop.slotName,
                                         prop.name

--- a/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/ImplGenerator.kt
+++ b/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/generator/ImplGenerator.kt
@@ -4,6 +4,7 @@ import com.squareup.kotlinpoet.*
 import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
 import org.babyfish.jimmer.ksp.immutable.meta.ImmutableProp
 import org.babyfish.jimmer.ksp.immutable.meta.ImmutableType
+import org.babyfish.jimmer.ksp.name
 import org.babyfish.jimmer.ksp.util.generatedAnnotation
 import org.babyfish.jimmer.meta.PropId
 import kotlin.reflect.KClass
@@ -191,7 +192,7 @@ class ImplGenerator(
                                             beginControlFlow(
                                                 when {
                                                     prop.loadedFieldName !== null -> "if (!${prop.loadedFieldName})"
-                                                    else -> "if (${prop.valueFieldName} === null)"
+                                                    else -> "if (${prop.valueFieldName} ${if (prop.isUnsigned) "==" else "==="} null)"
                                                 }
                                             )
                                             addStatement(
@@ -317,7 +318,7 @@ class ImplGenerator(
                                         unindent()
                                     }
                                     else -> {
-                                        val cond = prop.loadedFieldName ?: "${prop.valueFieldName} !== null"
+                                        val cond = prop.loadedFieldName ?: "${prop.valueFieldName} ${if (prop.isUnsigned) "!=" else "!=="} null"
                                         addStatement("%L", cond)
                                     }
                                 }
@@ -386,7 +387,7 @@ class ImplGenerator(
                                 }
                                 beginControlFlow(
                                     "if (%L)",
-                                    prop.loadedFieldName ?: "${prop.valueFieldName} !== null"
+                                    prop.loadedFieldName ?: "${prop.valueFieldName} ${if (prop.isUnsigned) "!=" else "!=="} null"
                                 )
                                 add("hash = 31 * hash + ")
                                 if (shallow && prop.isAssociation(false)) {
@@ -457,7 +458,7 @@ class ImplGenerator(
                                     continue
                                 }
                                 val localLoadedName = "__${prop.name}Loaded"
-                                val objLoadedName = prop.loadedFieldName ?: "${prop.valueFieldName} !== null"
+                                val objLoadedName = prop.loadedFieldName ?: "${prop.valueFieldName} ${if (prop.isUnsigned) "!=" else "!=="} null"
                                 add("val %L = \n", localLoadedName)
                                 addStatement("    this.%L", objLoadedName)
                                 beginControlFlow(

--- a/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/meta/ImmutableProp.kt
+++ b/project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/immutable/meta/ImmutableProp.kt
@@ -15,7 +15,6 @@ import org.babyfish.jimmer.dto.compiler.spi.BaseProp
 import org.babyfish.jimmer.impl.util.Keywords
 import org.babyfish.jimmer.jackson.JsonConverter
 import org.babyfish.jimmer.ksp.*
-import org.babyfish.jimmer.ksp.client.ClientProcessor.Companion.realDeclaration
 import org.babyfish.jimmer.ksp.immutable.generator.DRAFT
 import org.babyfish.jimmer.ksp.immutable.generator.KEY_FULL_NAME
 import org.babyfish.jimmer.ksp.immutable.generator.parseValidationMessages
@@ -161,6 +160,15 @@ class ImmutableProp(
 
     override val isNullable: Boolean
         get() = _isNullable
+
+    val isUnsigned: Boolean =
+        when (propDeclaration.type.resolve().declaration.qualifiedName!!.asString()) {
+            ULong::class.qualifiedName -> true
+            UInt::class.qualifiedName -> true
+            UShort::class.qualifiedName -> true
+            UByte::class.qualifiedName -> true
+            else -> false
+        }
 
     init {
         val descriptor = PropDescriptor

--- a/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/KSqlClient.kt
+++ b/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/KSqlClient.kt
@@ -23,6 +23,7 @@ import org.babyfish.jimmer.sql.kt.impl.KSqlClientImpl
 import org.babyfish.jimmer.sql.runtime.EntityManager
 import org.babyfish.jimmer.sql.runtime.Executor
 import org.babyfish.jimmer.sql.runtime.JSqlClientImplementor
+import org.babyfish.jimmer.sql.runtime.ScalarProvider
 import java.sql.Connection
 import kotlin.reflect.KClass
 import kotlin.reflect.KProperty1
@@ -1057,8 +1058,52 @@ interface KSqlClient {
     val javaClient: JSqlClientImplementor
 }
 
+internal class UnsignedLongScalar : ScalarProvider<ULong, String> {
+    override fun toScalar(sqlValue: String): ULong {
+        return sqlValue.toULong()
+    }
+
+    override fun toSql(scalarValue: ULong): String {
+        return scalarValue.toString()
+    }
+}
+
+internal class UnsignedIntScalar : ScalarProvider<UInt, String> {
+    override fun toScalar(sqlValue: String): UInt {
+        return sqlValue.toUInt()
+    }
+
+    override fun toSql(scalarValue: UInt): String {
+        return scalarValue.toString()
+    }
+}
+
+internal class UnsignedShortScalar : ScalarProvider<UShort, String> {
+    override fun toScalar(sqlValue: String): UShort {
+        return sqlValue.toUShort()
+    }
+
+    override fun toSql(scalarValue: UShort): String {
+        return scalarValue.toString()
+    }
+}
+
+internal class UnsignedByteScalar : ScalarProvider<UByte, String> {
+    override fun toScalar(sqlValue: String): UByte {
+        return sqlValue.toUByte()
+    }
+
+    override fun toSql(scalarValue: UByte): String {
+        return scalarValue.toString()
+    }
+}
+
 fun newKSqlClient(block: KSqlClientDsl.() -> Unit): KSqlClient {
     val javaBuilder = JSqlClient.newBuilder()
+    javaBuilder.addScalarProvider(UnsignedLongScalar())
+    javaBuilder.addScalarProvider(UnsignedIntScalar())
+    javaBuilder.addScalarProvider(UnsignedShortScalar())
+    javaBuilder.addScalarProvider(UnsignedByteScalar())
     val dsl = KSqlClientDsl(javaBuilder)
     dsl.block()
     return dsl.buildKSqlClient()

--- a/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/KExpressions.kt
+++ b/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/KExpressions.kt
@@ -355,7 +355,254 @@ operator fun <N: Number> KExpression<N>.rem(right: KExpression<N>): KNullableExp
 operator fun <N: Number> KExpression<N>.rem(right: N): KNullableExpression<N> =
     BinaryOperatorExpression.NullableRem(this, value(right))
 
+object UnsignedLong {
+    operator fun KNonNullExpression<ULong>.plus(right: KNonNullExpression<ULong>): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullPlus(this, right)
 
+    operator fun KNonNullExpression<ULong>.plus(right: ULong): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullPlus(this, value(right))
+
+    operator fun KExpression<ULong>.plus(right: KExpression<ULong>): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullablePlus(this, right)
+
+    operator fun KExpression<ULong>.plus(right: ULong): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullablePlus(this, value(right))
+
+    operator fun KNonNullExpression<ULong>.minus(right: KNonNullExpression<ULong>): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullMinus(this, right)
+
+    operator fun KNonNullExpression<ULong>.minus(right: ULong): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullMinus(this, value(right))
+
+    operator fun KExpression<ULong>.minus(right: KExpression<ULong>): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableMinus(this, right)
+
+    operator fun KExpression<ULong>.minus(right: ULong): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableMinus(this, value(right))
+
+    operator fun KNonNullExpression<ULong>.times(right: KNonNullExpression<ULong>): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullTimes(this, right)
+
+    operator fun KNonNullExpression<ULong>.times(right: ULong): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullTimes(this, value(right))
+
+    operator fun KExpression<ULong>.times(right: KExpression<ULong>): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableTimes(this, right)
+
+    operator fun KExpression<ULong>.times(right: ULong): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableTimes(this, value(right))
+
+    operator fun KNonNullExpression<ULong>.div(right: KNonNullExpression<ULong>): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullDiv(this, right)
+
+    operator fun KNonNullExpression<ULong>.div(right: ULong): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullDiv(this, value(right))
+
+    operator fun KExpression<ULong>.div(right: KExpression<ULong>): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableDiv(this, right)
+
+    operator fun KExpression<ULong>.div(right: ULong): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableDiv(this, value(right))
+
+    operator fun KNonNullExpression<ULong>.rem(right: KNonNullExpression<ULong>): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullRem(this, right)
+
+    operator fun KNonNullExpression<ULong>.rem(right: ULong): KNonNullExpression<ULong> =
+        BinaryOperatorExpression.NonNullRem(this, value(right))
+
+    operator fun KExpression<ULong>.rem(right: KExpression<ULong>): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableRem(this, right)
+
+    operator fun KExpression<ULong>.rem(right: ULong): KNullableExpression<ULong> =
+        BinaryOperatorExpression.NullableRem(this, value(right))
+}
+
+object UnsignedInt {
+    operator fun KNonNullExpression<UInt>.plus(right: KNonNullExpression<UInt>): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullPlus(this, right)
+
+    operator fun KNonNullExpression<UInt>.plus(right: UInt): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullPlus(this, value(right))
+
+    operator fun KExpression<UInt>.plus(right: KExpression<UInt>): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullablePlus(this, right)
+
+    operator fun KExpression<UInt>.plus(right: UInt): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullablePlus(this, value(right))
+
+    operator fun KNonNullExpression<UInt>.minus(right: KNonNullExpression<UInt>): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullMinus(this, right)
+
+    operator fun KNonNullExpression<UInt>.minus(right: UInt): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullMinus(this, value(right))
+
+    operator fun KExpression<UInt>.minus(right: KExpression<UInt>): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableMinus(this, right)
+
+    operator fun KExpression<UInt>.minus(right: UInt): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableMinus(this, value(right))
+
+    operator fun KNonNullExpression<UInt>.times(right: KNonNullExpression<UInt>): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullTimes(this, right)
+
+    operator fun KNonNullExpression<UInt>.times(right: UInt): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullTimes(this, value(right))
+
+    operator fun KExpression<UInt>.times(right: KExpression<UInt>): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableTimes(this, right)
+
+    operator fun KExpression<UInt>.times(right: UInt): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableTimes(this, value(right))
+
+    operator fun KNonNullExpression<UInt>.div(right: KNonNullExpression<UInt>): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullDiv(this, right)
+
+    operator fun KNonNullExpression<UInt>.div(right: UInt): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullDiv(this, value(right))
+
+    operator fun KExpression<UInt>.div(right: KExpression<UInt>): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableDiv(this, right)
+
+    operator fun KExpression<UInt>.div(right: UInt): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableDiv(this, value(right))
+
+    operator fun KNonNullExpression<UInt>.rem(right: KNonNullExpression<UInt>): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullRem(this, right)
+
+    operator fun KNonNullExpression<UInt>.rem(right: UInt): KNonNullExpression<UInt> =
+        BinaryOperatorExpression.NonNullRem(this, value(right))
+
+    operator fun KExpression<UInt>.rem(right: KExpression<UInt>): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableRem(this, right)
+
+    operator fun KExpression<UInt>.rem(right: UInt): KNullableExpression<UInt> =
+        BinaryOperatorExpression.NullableRem(this, value(right))
+
+}
+
+object UnsignedShort {
+    operator fun KNonNullExpression<UShort>.plus(right: KNonNullExpression<UShort>): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullPlus(this, right)
+
+    operator fun KNonNullExpression<UShort>.plus(right: UShort): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullPlus(this, value(right))
+
+    operator fun KExpression<UShort>.plus(right: KExpression<UShort>): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullablePlus(this, right)
+
+    operator fun KExpression<UShort>.plus(right: UShort): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullablePlus(this, value(right))
+
+    operator fun KNonNullExpression<UShort>.minus(right: KNonNullExpression<UShort>): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullMinus(this, right)
+
+    operator fun KNonNullExpression<UShort>.minus(right: UShort): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullMinus(this, value(right))
+
+    operator fun KExpression<UShort>.minus(right: KExpression<UShort>): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableMinus(this, right)
+
+    operator fun KExpression<UShort>.minus(right: UShort): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableMinus(this, value(right))
+
+    operator fun KNonNullExpression<UShort>.times(right: KNonNullExpression<UShort>): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullTimes(this, right)
+
+    operator fun KNonNullExpression<UShort>.times(right: UShort): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullTimes(this, value(right))
+
+    operator fun KExpression<UShort>.times(right: KExpression<UShort>): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableTimes(this, right)
+
+    operator fun KExpression<UShort>.times(right: UShort): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableTimes(this, value(right))
+
+    operator fun KNonNullExpression<UShort>.div(right: KNonNullExpression<UShort>): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullDiv(this, right)
+
+    operator fun KNonNullExpression<UShort>.div(right: UShort): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullDiv(this, value(right))
+
+    operator fun KExpression<UShort>.div(right: KExpression<UShort>): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableDiv(this, right)
+
+    operator fun KExpression<UShort>.div(right: UShort): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableDiv(this, value(right))
+
+    operator fun KNonNullExpression<UShort>.rem(right: KNonNullExpression<UShort>): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullRem(this, right)
+
+    operator fun KNonNullExpression<UShort>.rem(right: UShort): KNonNullExpression<UShort> =
+        BinaryOperatorExpression.NonNullRem(this, value(right))
+
+    operator fun KExpression<UShort>.rem(right: KExpression<UShort>): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableRem(this, right)
+
+    operator fun KExpression<UShort>.rem(right: UShort): KNullableExpression<UShort> =
+        BinaryOperatorExpression.NullableRem(this, value(right))
+}
+
+object UnsignedByte {
+    operator fun KNonNullExpression<UByte>.plus(right: KNonNullExpression<UByte>): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullPlus(this, right)
+
+    operator fun KNonNullExpression<UByte>.plus(right: UByte): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullPlus(this, value(right))
+
+    operator fun KExpression<UByte>.plus(right: KExpression<UByte>): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullablePlus(this, right)
+
+    operator fun KExpression<UByte>.plus(right: UByte): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullablePlus(this, value(right))
+
+    operator fun KNonNullExpression<UByte>.minus(right: KNonNullExpression<UByte>): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullMinus(this, right)
+
+    operator fun KNonNullExpression<UByte>.minus(right: UByte): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullMinus(this, value(right))
+
+    operator fun KExpression<UByte>.minus(right: KExpression<UByte>): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableMinus(this, right)
+
+    operator fun KExpression<UByte>.minus(right: UByte): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableMinus(this, value(right))
+
+    operator fun KNonNullExpression<UByte>.times(right: KNonNullExpression<UByte>): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullTimes(this, right)
+
+    operator fun KNonNullExpression<UByte>.times(right: UByte): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullTimes(this, value(right))
+
+    operator fun KExpression<UByte>.times(right: KExpression<UByte>): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableTimes(this, right)
+
+    operator fun KExpression<UByte>.times(right: UByte): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableTimes(this, value(right))
+
+    operator fun KNonNullExpression<UByte>.div(right: KNonNullExpression<UByte>): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullDiv(this, right)
+
+    operator fun KNonNullExpression<UByte>.div(right: UByte): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullDiv(this, value(right))
+
+    operator fun KExpression<UByte>.div(right: KExpression<UByte>): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableDiv(this, right)
+
+    operator fun KExpression<UByte>.div(right: UByte): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableDiv(this, value(right))
+
+    operator fun KNonNullExpression<UByte>.rem(right: KNonNullExpression<UByte>): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullRem(this, right)
+
+    operator fun KNonNullExpression<UByte>.rem(right: UByte): KNonNullExpression<UByte> =
+        BinaryOperatorExpression.NonNullRem(this, value(right))
+
+    operator fun KExpression<UByte>.rem(right: KExpression<UByte>): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableRem(this, right)
+
+    operator fun KExpression<UByte>.rem(right: UByte): KNullableExpression<UByte> =
+        BinaryOperatorExpression.NullableRem(this, value(right))
+}
 
 operator fun <N: Number> KNonNullExpression<N>.unaryPlus() : KNonNullExpression<N> =
     this

--- a/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/impl/BinaryOperatorExpression.kt
+++ b/project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/impl/BinaryOperatorExpression.kt
@@ -6,7 +6,7 @@ import org.babyfish.jimmer.sql.kt.ast.expression.KExpression
 import org.babyfish.jimmer.sql.kt.ast.expression.KNonNullExpression
 import org.babyfish.jimmer.sql.kt.ast.expression.KNullableExpression
 
-internal abstract class BinaryOperatorExpression<N: Number>(
+internal abstract class BinaryOperatorExpression<N: Any>(
     private var left: KExpression<N>,
     private var right: KExpression<N>
 ) : AbstractKExpression<N>() {
@@ -39,7 +39,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
 
     protected abstract fun operator(): String
 
-    class NonNullPlus<N: Number>(
+    class NonNullPlus<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNonNullExpression<N> {
@@ -49,7 +49,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.PLUS
     }
 
-    class NonNullMinus<N: Number>(
+    class NonNullMinus<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNonNullExpression<N> {
@@ -59,7 +59,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.PLUS
     }
 
-    class NonNullTimes<N: Number>(
+    class NonNullTimes<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNonNullExpression<N> {
@@ -69,7 +69,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.TIMES
     }
 
-    class NonNullDiv<N: Number>(
+    class NonNullDiv<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNonNullExpression<N> {
@@ -79,7 +79,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.TIMES
     }
 
-    class NonNullRem<N: Number>(
+    class NonNullRem<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNonNullExpression<N> {
@@ -89,7 +89,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.TIMES
     }
 
-    class NullablePlus<N: Number>(
+    class NullablePlus<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNullableExpression<N> {
@@ -99,7 +99,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.PLUS
     }
 
-    class NullableMinus<N: Number>(
+    class NullableMinus<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNullableExpression<N> {
@@ -109,7 +109,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.PLUS
     }
 
-    class NullableTimes<N: Number>(
+    class NullableTimes<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNullableExpression<N> {
@@ -119,7 +119,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.TIMES
     }
 
-    class NullableDiv<N: Number>(
+    class NullableDiv<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNullableExpression<N> {
@@ -129,7 +129,7 @@ internal abstract class BinaryOperatorExpression<N: Number>(
         override fun precedence(): Int = ExpressionPrecedences.TIMES
     }
 
-    class NullableRem<N: Number>(
+    class NullableRem<N: Any>(
         left: KExpression<N>,
         right: KExpression<N>
     ): BinaryOperatorExpression<N>(left, right), KNullableExpression<N> {

--- a/project/jimmer-sql-kotlin/src/test/dto/UnsignedNumber.dto
+++ b/project/jimmer-sql-kotlin/src/test/dto/UnsignedNumber.dto
@@ -1,0 +1,16 @@
+export org.babyfish.jimmer.sql.kt.model.UnsignedNumber
+    -> package org.babyfish.jimmer.sql.model.dto
+
+input UnsignedNumberInput {
+    id
+    unsignedInt
+    unsignedShort
+    unsignedByte
+}
+
+UnsignedNumberView {
+    id
+    unsignedInt
+    unsignedShort
+    unsignedByte
+}

--- a/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/common/AbstractTest.kt
+++ b/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/common/AbstractTest.kt
@@ -2,6 +2,7 @@ package org.babyfish.jimmer.sql.kt.common
 
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
+import com.fasterxml.jackson.module.kotlin.KotlinModule
 import org.babyfish.jimmer.jackson.ImmutableModule
 import org.babyfish.jimmer.meta.ImmutableProp
 import org.babyfish.jimmer.sql.ast.impl.mutation.QueryReason
@@ -169,6 +170,7 @@ abstract class AbstractTest {
         protected val MAPPER = ObjectMapper()
             .registerModule(ImmutableModule())
             .registerModule(JavaTimeModule())
+            .registerModule(KotlinModule.Builder().build())
 
         @JvmStatic
         protected fun contentEquals(

--- a/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/meta/UnsignedNumberTest.kt
+++ b/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/meta/UnsignedNumberTest.kt
@@ -1,0 +1,115 @@
+package org.babyfish.jimmer.sql.kt.meta
+
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedByte.div
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedByte.minus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedByte.plus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedByte.times
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedInt.div
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedInt.minus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedInt.plus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedInt.times
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedLong.div
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedLong.minus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedLong.plus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedLong.times
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedShort.div
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedShort.minus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedShort.plus
+import org.babyfish.jimmer.sql.kt.ast.expression.UnsignedShort.times
+import org.babyfish.jimmer.sql.kt.ast.expression.eq
+import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
+import org.babyfish.jimmer.sql.kt.common.assertContent
+import org.babyfish.jimmer.sql.kt.model.*
+import org.babyfish.jimmer.sql.model.dto.UnsignedNumberInput
+import org.babyfish.jimmer.sql.model.dto.UnsignedNumberView
+import kotlin.test.Test
+
+class UnsignedNumberTest : AbstractQueryTest() {
+    @Test
+    fun test() {
+        UnsignedNumber {
+            id = 1u
+            unsignedInt = 1u
+            unsignedShort = 1u
+            unsignedByte = 1u
+        }
+    }
+
+    @Test
+    fun testInput() {
+        assertContent(
+            "{\"id\":1,\"unsignedInt\":1,\"unsignedShort\":1,\"unsignedByte\":1}",
+            UnsignedNumberInput(
+                id = 1u,
+                unsignedInt = 1u,
+                unsignedShort = 1u,
+                unsignedByte = 1u
+            ).toEntity()
+        )
+    }
+
+    @Test
+    fun testQueryTuple() {
+        executeAndExpect(
+            sqlClient.createQuery(UnsignedNumber::class) {
+                where(table.id eq 1u)
+                select(
+                    table.id.plus(1u).minus(1u).times(2u).div(2u),
+                    table.unsignedInt.plus(1u).minus(1u).times(2u).div(2u),
+                    (table.unsignedShort + 1u - 1u) * 2u / 2u,
+                    (table.unsignedByte + 1u - 1u) * 2u / 2u
+                )
+            }
+        ) {
+            statement(0)
+                .sql("select (tb_1_.ID + ? - ?) * ? / ?, (tb_1_.UNSIGNED_INT + ? - ?) * ? / ?, (tb_1_.UNSIGNED_SHORT + ? - ?) * ? / ?, (tb_1_.UNSIGNED_BYTE + ? - ?) * ? / ? from UNSIGNED_NUMBER tb_1_ where tb_1_.ID = ?")
+            rows("[{\"_1\":1,\"_2\":1,\"_3\":1,\"_4\":1}]")
+        }
+    }
+
+    @Test
+    fun testQueryFetch() {
+        executeAndExpect(
+            sqlClient.createQuery(UnsignedNumber::class) {
+                where(table.id eq 1u)
+                select(table.fetchBy {
+                    unsignedInt()
+                    unsignedShort()
+                    unsignedByte()
+                })
+            }
+        ) {
+            statement(0)
+                .sql("select tb_1_.ID, tb_1_.UNSIGNED_INT, tb_1_.UNSIGNED_SHORT, tb_1_.UNSIGNED_BYTE from UNSIGNED_NUMBER tb_1_ where tb_1_.ID = ?")
+            rows("[{\"id\":1,\"unsignedInt\":1,\"unsignedShort\":1,\"unsignedByte\":1}]")
+        }
+    }
+
+    @Test
+    fun testQueryView() {
+        executeAndExpect(
+            sqlClient.createQuery(UnsignedNumber::class) {
+                where(table.id eq 1u)
+                select(table.fetch(UnsignedNumberView::class))
+            }
+        ) {
+            statement(0)
+                .sql("select tb_1_.ID, tb_1_.UNSIGNED_INT, tb_1_.UNSIGNED_SHORT, tb_1_.UNSIGNED_BYTE from UNSIGNED_NUMBER tb_1_ where tb_1_.ID = ?")
+            rows("[{\"id\":1,\"unsignedInt\":1,\"unsignedShort\":1,\"unsignedByte\":1}]")
+        }
+    }
+
+    @Test
+    fun testQuery() {
+        executeAndExpect(
+            sqlClient.createQuery(UnsignedNumber::class) {
+                where(table.id eq 1u)
+                select(table)
+            }
+        ) {
+            statement(0)
+                .sql("select tb_1_.ID, tb_1_.UNSIGNED_INT, tb_1_.UNSIGNED_SHORT, tb_1_.UNSIGNED_BYTE from UNSIGNED_NUMBER tb_1_ where tb_1_.ID = ?")
+            rows("[{\"id\":1,\"unsignedInt\":1,\"unsignedShort\":1,\"unsignedByte\":1}]")
+        }
+    }
+}

--- a/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/UnsignedNumber.kt
+++ b/project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/UnsignedNumber.kt
@@ -1,0 +1,16 @@
+package org.babyfish.jimmer.sql.kt.model
+
+import org.babyfish.jimmer.sql.Entity
+import org.babyfish.jimmer.sql.Id
+
+@Entity
+interface UnsignedNumber {
+    @Id
+    val id: ULong
+
+    val unsignedInt: UInt
+
+    val unsignedShort: UShort
+
+    val unsignedByte: UByte
+}

--- a/project/jimmer-sql-kotlin/src/test/resources/database.sql
+++ b/project/jimmer-sql-kotlin/src/test/resources/database.sql
@@ -28,6 +28,7 @@ drop table book_store if exists;
 drop table tree_node if exists;
 drop table primitive if exists;
 drop table personal if exists;
+drop table unsigned_number if exists;
 
 drop sequence file_user_id_seq if exists;
 drop sequence file_id_seq if exists;
@@ -762,4 +763,18 @@ alter table personal
         primary key(id);
 
 insert into personal(id, phone) values
-    (1, '12345678910')
+    (1, '12345678910');
+
+create table unsigned_number(
+    id bigint not null,
+    unsigned_int int not null,
+    unsigned_short smallint not null,
+    unsigned_byte tinyint not null
+);
+
+alter table unsigned_number
+    add constraint pk_unsigned_number
+        primary key(id);
+
+insert into unsigned_number(id, unsigned_int, unsigned_short, unsigned_byte) values
+    (1, 1, 1, 1);

--- a/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/association/meta/AssociationProp.java
+++ b/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/association/meta/AssociationProp.java
@@ -1,5 +1,9 @@
 package org.babyfish.jimmer.sql.association.meta;
 
+import kotlin.UByte;
+import kotlin.UInt;
+import kotlin.ULong;
+import kotlin.UShort;
 import org.babyfish.jimmer.jackson.Converter;
 import org.babyfish.jimmer.jackson.ConverterMetadata;
 import org.babyfish.jimmer.lang.Ref;

--- a/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/runtime/DefaultExecutor.java
+++ b/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/runtime/DefaultExecutor.java
@@ -1,5 +1,9 @@
 package org.babyfish.jimmer.sql.runtime;
 
+import kotlin.UByte;
+import kotlin.UInt;
+import kotlin.ULong;
+import kotlin.UShort;
 import org.babyfish.jimmer.impl.util.Classes;
 import org.babyfish.jimmer.meta.ImmutableProp;
 import org.babyfish.jimmer.sql.collection.TypedList;
@@ -44,6 +48,18 @@ public class DefaultExecutor implements Executor {
                         parameterIndex.get(),
                         stmt.getConnection().createArrayOf(typedList.getSqlElementType(), typedList.toArray())
                 );
+            } else if (variable.getClass().getName().startsWith("kotlin.")) {
+                final String name = variable.getClass().getName();
+                if (name.equals(ULong.class.getName())) {
+                    variable = Long.parseLong(variable.toString());
+                } else if (name.equals(UInt.class.getName())) {
+                    variable = Integer.parseInt(variable.toString());
+                } else if (name.equals(UShort.class.getName())) {
+                    variable = Short.parseShort(variable.toString());
+                } else if (name.equals(UByte.class.getName())) {
+                    variable = Byte.parseByte(variable.toString());
+                }
+                stmt.setObject(parameterIndex.get(), variable);
             } else {
                 stmt.setObject(parameterIndex.get(), variable);
             }

--- a/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/runtime/ReaderManager.java
+++ b/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/runtime/ReaderManager.java
@@ -15,10 +15,8 @@ import org.babyfish.jimmer.sql.association.Association;
 import org.babyfish.jimmer.sql.association.meta.AssociationType;
 import org.babyfish.jimmer.sql.dialect.Dialect;
 import org.babyfish.jimmer.sql.exception.ExecutionException;
-import org.babyfish.jimmer.sql.meta.ColumnDefinition;
 import org.babyfish.jimmer.sql.meta.FormulaTemplate;
 import org.babyfish.jimmer.sql.meta.SqlTemplate;
-import org.babyfish.jimmer.sql.meta.Storage;
 import org.jetbrains.annotations.NotNull;
 
 import java.lang.reflect.ParameterizedType;
@@ -125,6 +123,9 @@ public class ReaderManager {
         ScalarProvider<Object, Object> scalarProvider = sqlClient.getScalarProvider(prop);
         if (scalarProvider != null) {
             return scalarProviderReader(scalarProvider);
+        }
+        if (prop.isUnsigned()) {
+            return scalarReader(prop.getElementClass());
         }
         if (sqlClient.getDialect().isArraySupported()) {
             Class<?> returnClass = prop.getReturnClass();

